specs/virtual_state/end_tx.md

```diff
@@ -2,67 +2,51 @@
 
 ## Procedure
 
-The `end_tx` gadget performs preliminary checks on a transaction and its caller's account as well as further checks depending on the transaction type. 
+The `end_tx` gadget handles refund and coinbase tip as well as cumulative gas within a block. 
 
-More particularly, for all transactions, we make sure that the account nonce is incremented by 1 if the tx is valid, and that the caller address is not null. We also verify that a transaction is set as invalid if and only if there is either not enough gas, the tx's nonce is not the same as caller's account previous nonce or the balance is insufficient.
-
-Depending whether the transaction is a Create, calls a precompile or calls an account with or without code, we do the following,
-- isCreate = 1
-- calling Precompile
-- Call to account with empty code
-- Call to account with code
 
 ## Constraints
 
-1. If the transaction is invalid, check the refund is null
--> `(1 - is_tx_invalid) * effective_refund = 0`
-2. Add effective_refund * gas_price back to caller's balance
-3. Add gas_used * effective_tip to coinbase's balance
-4. Constrain tx status matches with `PostStateOrStatus`
-5. Constrain log id matches with `LogLength` of TxReceipt tag in RW
-6. If tx is invalid, assert log_id is 0
-7. Constrain `CumulativeGasUsed` of TxReceipt tag in RW to 0 if is first tx otherwise previous `CumulativeGasUsed + gas_used`
-8. if next execution state is begin_tx
-    1. Assert tx_id is incremented by 1
-    2. Constrain state transition with rw_counter =  10 - is_first_tx
-9. if next execution step is EndBlock
-    1. Constrain state transition with rw_counter =  9 - is_first_tx and same call_id
+1. Calculate effective refund (c.f. below list)
+2. Add [(effective_refund + gas_left) * gas_price] back to caller's balance
+3. Check effective tip = tx gas price - base fee
+4. Add gas_used * effective_tip to coinbase's balance (and create coinbase account if needs be)
+5. End transaction
+    1. Check that the current cumulative gas used is null if this is the 1st transaction else corresponds to previous transaction's
+    2. Increase cumulative gas used by the gas used
+    3. If the next execution state is BeginTx
+        1. The next transaction's id is correctly incremented
+        2. Transition to new context (rw_counter =  8 + transfer.rw + is_first_tx)
```

Comment:
From the implementation, rw_counter should be "8 + transfer.rw `-` is_first_tx".
It's the same in 4.1 below.

Reply:
I have issues with this particular lookup it seems 🤦

Reply:
(so it is 9 + transfer.rw - is_first_tx)

Reply:
(I checked thrice)